src/solc_0.8/defi/rules/ContributionRules.sol

```diff
@@ -0,0 +1,194 @@
+//SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.2;
+
+import {Ownable} from "@openzeppelin/contracts-0.8/access/Ownable.sol";
+import {Address} from "@openzeppelin/contracts-0.8/utils/Address.sol";
+import {SafeMathWithRequire} from "../../common/Libraries/SafeMathWithRequire.sol";
+import {IERC721} from "@openzeppelin/contracts-0.8/token/ERC721/IERC721.sol";
+import {IERC1155} from "@openzeppelin/contracts-0.8/token/ERC1155/IERC1155.sol";
+
+contract ContributionRules is Ownable {
+    using Address for address;
+    uint256 internal constant DECIMALS_7 = 10000000;
+    uint256 internal constant MIDPOINT_9 = 500000000;
+    uint256 internal constant NFT_FACTOR_6 = 10000;
+    uint256 internal constant NFT_CONSTANT_3 = 9000;
+    uint256 internal constant ROOT3_FACTOR = 697;
+
+    struct MultiplierERC721 {
+        uint256[] ids;
+        uint256[] multiplier;
+        uint256 index;
+        bool balanceOf;
+    }
+
+    struct MultiplierERC1155 {
+        IERC1155 assetContract;
+        uint256[] ids;
+        uint256[] multiplier;
+        uint256 index;
+    }
+
+    mapping(IERC721 => MultiplierERC721) private _listERC721;
+    mapping(IERC1155 => MultiplierERC1155) private _listERC1155;
+    IERC721[] internal _listERC721Index;
+    IERC1155[] internal _listERC1155Index;
+
+    IERC721 public landMultiplierContract;
+
+    constructor(IERC721 _landMultiplierContract) {
+        landMultiplierContract = _landMultiplierContract;
+    }
+
+    function computeMultiplier(address account, uint256 amountStaked) external view returns (uint256) {
+        uint256 nftMultiplier = _multiplierBalanceOfERC721(account);
+        uint256 assetMultiplier = _multiplierBalanceOfERC1155(account);
+        uint256 landMultiplier = _multiplierBalanceOfLand(account);
+
+        return amountStaked + ((amountStaked * (nftMultiplier + assetMultiplier + landMultiplier)) / 100);
+    }
+
+    function setLandMultiplierContract(address newLandContract) external onlyOwner {
+        require(newLandContract.isContract(), "ContributionRules: invalid contract address");
+        landMultiplierContract = IERC721(newLandContract);
+    }
+
+    function addERC1155MultiplierList(
+        address contractERC1155,
+        uint256[] memory ids,
+        uint256[] memory multiplier
+    ) external onlyOwner {
+        require(contractERC1155.isContract(), "ContributionRules: invalid contract address");
+        require(ids.length > 0, "ContributionRules: invalid ids array");
+        require(multiplier.length > 0, "ContributionRules: invalid multiplier array");
+
+        IERC1155 multContract = IERC1155(contractERC1155);
+
+        _listERC1155[multContract].ids = ids;
+        _listERC1155[multContract].multiplier = multiplier;
+
+        if (_islistERC1155Member(multContract) == false) {
+            _listERC1155Index.push(multContract);
+            _listERC1155[multContract].index = _listERC1155Index.length - 1;
+        }
```

Comment:
else revert ?

Reply:
we gonna name it set, as it add and update